cmd/clair/main.go

```diff
@@ -51,54 +56,70 @@ func main() {
 		golog.Fatalf("failed to validate config: %v", err)
 	}
 
-	// initialize performs all Clair initialization tasks.
-	init, err := initialize.New(conf)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ctx, err = initialize.Logging(ctx, &conf)
 	if err != nil {
-		golog.Fatalf("initialized failed: %v", err)
+		golog.Fatalf("failed to set up logging: %v", err)
 	}
-	logger := zerolog.Ctx(init.GlobalCTX).With().Str("component", "main").Logger()
+	logger := zerolog.Ctx(ctx).With().Str("component", "main").Logger()
+	logger.Info().Str("version", Version).Msg("starting")
 
-	// register signal handler
-	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt)
-	logger.Info().Msg("registered signal handler")
+	// Some machinery for starting and stopping server goroutines
+	down := &Shutdown{}
+	srvs, srvctx := errgroup.WithContext(ctx)
 
-	// introspection server
-	if init.Introspection != nil {
+	srvs.Go(func() (_ error) {
 		logger.Info().Msg("launching introspection server")
-		go func() {
-			err := init.Introspection.ListenAndServe()
-			if err != nil {
-				logger.Err(err).Msg("introspection server failed to launch. continuing anyway")
-			}
-		}()
-	}
+		i, err := introspection.New(ctx, conf, nil)
+		if err != nil {
+			logger.Warn().
+				Err(err).Msg("introspection server configuration failed. continuing anyway")
+			return
+		}
+		down.Add("introspection", i.Server)
+		if err := i.ListenAndServe(); err != http.ErrServerClosed {
```

Comment:
You've added the server to the Shutdown structure, but it can still fail to ListenAndServe. 

Is it okay for the Shutdown structure to call a server's "Shutdown" method if that server is not running? 

If not you - make sure it does not get into the Shutdown's structure's map. If it is okay, ignore this comment.

Reply:
I think this is okay, but I'm going to try to prove it to myself.

Reply:
https://play.golang.org/p/IbPqyX2MyY-

The playground seems broken right now, but running this locally doesn't panic.